service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/disk/GetAzureDiskAttachedVmStep.java

```diff
@@ -0,0 +1,83 @@
+import bio.terra.workspace.service.workspace.model.AzureCloudContext;
+import com.azure.core.management.exception.ManagementException;
+import com.azure.resourcemanager.compute.ComputeManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This steps works as a companion step for a disk deletion process. It gets identifier of a virtual
+ * machine to which data disk is attached and saves it into flight's working map. This data is
+ * required for 'undo' operation at DetachAzureDiskFromVmStep. It is required in case a disk has
+ * been detached from a virtual machine, but we need to attach it back in corresponding 'undo'
+ * operation.
+ */
+public class GetAzureDiskAttachedVmStep extends DeleteAzureControlledResourceStep {
+
+  private static final Logger logger = LoggerFactory.getLogger(GetAzureDiskAttachedVmStep.class);
+
+  private final CrlService crlService;
+  private final AzureConfiguration azureConfig;
+  private final ControlledAzureDiskResource resource;
+
+  public GetAzureDiskAttachedVmStep(
+      AzureConfiguration azureConfig, CrlService crlService, ControlledAzureDiskResource resource) {
+    this.crlService = crlService;
+    this.azureConfig = azureConfig;
+    this.resource = resource;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    final AzureCloudContext azureCloudContext =
+        DeleteAzureDiskFlightUtils.getAzureCloudContext(context);
+    final String diskResourceId =
+        DeleteAzureDiskFlightUtils.getAzureDiskResourceId(azureCloudContext, resource);
+    ComputeManager computeManager = crlService.getComputeManager(azureCloudContext, azureConfig);
+    try {
+      var disk = computeManager.disks().getById(diskResourceId);
+      if (disk.isAttachedToVirtualMachine()) {
+        context
+            .getWorkingMap()
+            .put(DeleteAzureDiskFlightUtils.DISK_ATTACHED_VM_ID_KEY, disk.virtualMachineId());
+      }
+    } catch (ManagementException ex) {
+      if (AzureManagementExceptionUtils.isExceptionCode(
+          ex, AzureManagementExceptionUtils.RESOURCE_NOT_FOUND)) {
+        logger.info(
+            "Disk '{}' does not exist in workspace {}.",
+            resource.getResourceId(),
+            resource.getWorkspaceId());
+        return StepResult.getStepResultSuccess();
+      }
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, ex);
+    }
+    return StepResult.getStepResultSuccess();
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    // nothing to undo here
+    return StepResult.getStepResultSuccess();
+  }
+
+  @Override
+  protected StepResult deleteResource(FlightContext context) throws InterruptedException {
+    /*
+    This step follows the contract for a step which deletes an Azure resource, but the implementation diverges from "standard".
```

Comment:
If this step isn't using _any_ of the functionality of `DeleteAzureControlledResourceStep`, I think it would probably be better to just use the `DeleteControlledResourceStep`, since that was intended to be a basic indicator interface anyway. 
If we think there's a way to improve `DeleteAzureControlledResourceStep` that will make it more applicable here, we should do that. 
We might also consider making `DeleteAzureControlledResourceStep.doStep()` final, just so it's clear what the pattern is.

I realize other opinions on this may vary - I'm not really intending to ask for specific changes here, just to prompt some examination and debate.

Reply:
The same here. I indented to use interface.